src/ints/int10_modes.cpp

```diff
@@ -2119,21 +2120,18 @@ std::optional<Rgb666> parse_color_token(const std::string& token,
 		}
 
 		if (is_hex3_token) {
-			auto r = static_cast<uint8_t>(value >> 8 & 0xf);
-			auto g = static_cast<uint8_t>(value >> 4 & 0xf);
-			auto b = static_cast<uint8_t>(value      & 0xf);
+			auto r4 = static_cast<uint8_t>(value >> 8 & 0xf);
+			auto g4 = static_cast<uint8_t>(value >> 4 & 0xf);
+			auto b4 = static_cast<uint8_t>(value      & 0xf);
 
-			r = r | r << 4;
-			g = g | g << 4;
-			b = b | b << 4;
+			return Rgb888::FromRgb444(r4, g4, b4);
 
-			return Rgb666(r, g, b);
-		} else {
-			auto r = static_cast<uint8_t>(value >> 16 & 0xff);
-			auto g = static_cast<uint8_t>(value >>  8 & 0xff);
-			auto b = static_cast<uint8_t>(value       & 0xff);
+		} else { // hex6 token
+			const auto r8 = static_cast<uint8_t>(value >> 16 & 0xff);
+			const auto g8 = static_cast<uint8_t>(value >>  8 & 0xff);
+			const auto b8 = static_cast<uint8_t>(value       & 0xff);
```

Comment:
Oops, resolved by mistake 😄 Yeah some of that colour parsing stuff does not belong here and should be extracted. The parsing of `cga_colors` will eventually go away; there's a ticket to turn those into resource files. So I'm leaving the parser bits that will go away anyway here for now.

Reply:
Actually, it's too tied to this whole `cga_colors` thing so I'll leave it be for now. It will be addressed in https://github.com/dosbox-staging/dosbox-staging/issues/2051 (made extra comment there about potentially extracting the hex colour parsing).